var/spack/repos/builtin/packages/py-darshan/package.py

```diff
@@ -0,0 +1,27 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyDarshan(PythonPackage):
+    """Python utilities to interact with Darshan log records of HPC applications."""
+
+    homepage = "https://www.mcs.anl.gov/research/projects/darshan"
+    pypi = "darshan/darshan-3.4.0.1.tar.gz"
+
+    maintainers = ["jeanbez", "shanedsnyder"]
+
+    version("3.4.0.1", sha256="0142fc7c0b12a9e5c22358aa26cca7083d28af42aeea7dfcc5698c56b6aee6b7")
+
+    depends_on("python@3.6:", type=("build", "run"))
+    depends_on("py-cffi", type=("build", "run"))
+    depends_on("py-numpy", type=("build", "run"))
+    depends_on("py-pandas", type=("build", "run"))
+    depends_on("py-matplotlib", type=("build", "run"))
+    depends_on("py-seaborn", type=("build", "run"))
+    depends_on("py-mako", type=("build", "run"))
+
+    depends_on("darshan-util", type=("build", "run"))
```

Comment:
I'm a bit confused about what is going on here. There's no `shared` variant of `darshan-util` spackage at this time, so I'm not clear on how you'll be able to link the built C object into the Python library?

You may find some related challenges/complications discussed at previous attempts to provide a draft version of a spackage:

- shared variant was drafted in here when doing some rough work: https://github.com/tylerjereddy/spack/pull/3
- original issue: https://github.com/darshan-hpc/darshan/issues/597

Did you actually test the build from source in a clean `spack` environment?

Reply:
Well @shanedsnyder seems to find that the new version of the C lib build system *does* produce the shared lib by default, so *maybe* you're ok here. Probably good to double check that `pytest --pyargs darshan` passes though, if you didn't already.

Reply:
Yep, `darshan-util` should be installing shared libs if `libtool` determines that the compiler can produce shared libs at configure time.

I suppose some spack compiler variants could not support shared libs -- is that something that could be handled gracefully somehow (i.e., pydarshan package can not be built if there is no `libdarshan-util.so`)? Otherwise, I assume the build of py-darshan will ultimately just fail.

Reply:
I checked that the tests don't pass though, that's always important to check.

- `spack install py-darshan % gcc@11.2.0 ^ python@3.10.4`
- `spack install py-pytest ^ python@3.10.4`
- `pytest --pyargs darshan`

```
================================================================================================================================ test session starts =================================================================================================================================
platform linux -- Python 3.10.4, pytest-7.1.3, pluggy-1.0.0
rootdir: /home/tyler/github_projects/spack, configfile: pytest.ini
collected 378 items / 2 errors                                                                                                                                                                                                                                                       

======================================================================================================================================= ERRORS =======================================================================================================================================
_____________________________________________ ERROR collecting opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/py-darshan-3.4.0.1-4dszt7wqbohm6yovl3pflwnizkrxqywh/lib/python3.10/site-packages/darshan/examples/tutorial/hello.py _____________________________________________
opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/py-darshan-3.4.0.1-4dszt7wqbohm6yovl3pflwnizkrxqywh/lib/python3.10/site-packages/darshan/examples/tutorial/hello.py:5: in <module>
    r = darshan.DarshanReport("ior_hdf5_example.darshan")
opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/py-darshan-3.4.0.1-4dszt7wqbohm6yovl3pflwnizkrxqywh/lib/python3.10/site-packages/darshan/report.py:374: in __init__
    self.open(filename, read_all=read_all)
opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/py-darshan-3.4.0.1-4dszt7wqbohm6yovl3pflwnizkrxqywh/lib/python3.10/site-packages/darshan/report.py:426: in open
    raise RuntimeError("Failed to open file.")
E   RuntimeError: Failed to open file.
---------------------------------------------------------------------------------------------------------------------------------- Captured stderr -----------------------------------------------------------------------------------------------------------------------------------
Error: darshan_log_open failed to open darshan log file ior_hdf5_example.darshan: No such file or directory.
_____________________________________________ ERROR collecting opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/py-darshan-3.4.0.1-4dszt7wqbohm6yovl3pflwnizkrxqywh/lib/python3.10/site-packages/darshan/examples/tutorial/plot.py ______________________________________________
ImportError while importing test module '/home/tyler/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/py-darshan-3.4.0.1-4dszt7wqbohm6yovl3pflwnizkrxqywh/lib/python3.10/site-packages/darshan/examples/tutorial/plot.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/python-3.10.4-syf6nm32j5ogompuqj5vdaxbxh3cve5p/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/py-darshan-3.4.0.1-4dszt7wqbohm6yovl3pflwnizkrxqywh/lib/python3.10/site-packages/darshan/examples/tutorial/plot.py:4: in <module>
    from darshan.experimental.plots.matplotlib import *
E   ModuleNotFoundError: No module named 'darshan.experimental.plots.matplotlib'
============================================================================================================================== short test summary info ===============================================================================================================================
ERROR opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/py-darshan-3.4.0.1-4dszt7wqbohm6yovl3pflwnizkrxqywh/lib/python3.10/site-packages/darshan/examples/tutorial/hello.py - RuntimeError: Failed to open file.
ERROR opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/py-darshan-3.4.0.1-4dszt7wqbohm6yovl3pflwnizkrxqywh/lib/python3.10/site-packages/darshan/examples/tutorial/plot.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================================================================================= 2 errors in 7.02s ==================================================================================================================================

```

Reply:
@tylerjereddy I did try running a clean install with an application that uses the library, and it works ok. I believe there might be something wrong with the tests, as it does not find the sample file. @shanedsnyder can you provide that file?

Reply:
The files should be installed as part of the installation, could be a difference on `spack` build system details perhaps. Either way, that's the normal incantation of the test suite which passes in our CI.

Reply:
perhaps stuff like `include_package_data` or `package_data` behavior has to be checked in the build logs when using `spack` & if there are any shims related to that stuff in standard `spack` installs

Reply:
One other thing that might help--it looks to me like the `*.darshan` files are installed properly by `spack` so they should be availble for testing. The traceback above shows attempts to import files that aren't under test, so maybe Adam might know why they would get imported by `spack` but not in conventional usage outside of `spack`.

Reply:
Spack isn't importing anything (at least in this test), the files picked up by pytest are. Spack has no fancy shims for package_data, we just run `pip install .` with a few extra flags to prevent downloads.

Reply:
The errors I report above are a result of running `pytest --pyargs darshan` in the `git` checkout of `spack` for this PR. The reason that happens is that `spack` has `pytest.ini` in its root with `python_files = *.py`, which overrides the largely-sensible default (https://docs.pytest.org/en/7.1.x/reference/reference.html#confval-python_files) of:

> files matching test_*.py and *_test.py will be considered test modules.

I won't comment on that--I assume the `spack` team has good reason to treat every single `*.py` file as under test/import by `pytest`, as surprising at it may be under normal circumstances.

Okay, so moving on from that, go to another directory like `/tmp` and run `pytest --pyargs darshan` on latest version of this branch:

`31 failed, 287 passed, 60 skipped, 76 warnings in 67.12s (0:01:07)`

All of those failures are real and informative, and could be easily missed if you only just ran an example workflow instead of the full suite. There are three reasons for failure:

1. `py-scipy` is a real (runtime) dependency of the project because a subset of codepaths/analysis pathways use an interpolation routine in `pandas` that needs SciPy--it turns out we did remove the need for SciPy more recently on `main`, but not in a tagged version yet, so I'd say you pretty much need to add it here for now.
2. `py-lxml` is mandatory to pass the tests only, because `pandas` is used to verify the integrity of some HTML report generation tests and it uses `lxml` for that read-in; Shane convinced me to make this an optional testing dependency in the future, and it is now in `main` where the relevant tests will skip, but until the next tagged release I'd say it may be sensible to add a dependency in the spackage, or at least a scoped test-stage dep.
3. `py-matplotlib` needs to be pinned or version capped at the tag of `py-darshan` in use here because of: https://github.com/darshan-hpc/darshan/issues/742 ; that will be relaxed in the next tagged release

So, adjusting based on the feedback from running the test suite, a diff like this, which is just a rough draft, gets us much closer:

```diff
diff --git a/var/spack/repos/builtin/packages/py-darshan/package.py b/var/spack/repos/builtin/packages/py-darshan/package.py
index 8f4202164a..4452453403 100644
--- a/var/spack/repos/builtin/packages/py-darshan/package.py
+++ b/var/spack/repos/builtin/packages/py-darshan/package.py
@@ -21,9 +21,12 @@ class PyDarshan(PythonPackage):
     depends_on("py-importlib-resources", when="^python@3.6", type=("build", "run"))
     depends_on("py-cffi", type=("build", "run"))
     depends_on("py-numpy@1.21:", type=("build", "run"))
+    depends_on("py-scipy", type=("run"))
     depends_on("py-pandas", type=("build", "run"))
-    depends_on("py-matplotlib", type=("build", "run"))
+    depends_on("py-matplotlib@3.4", type=("build", "run"))
     depends_on("py-seaborn", type=("build", "run"))
     depends_on("py-mako", type=("build", "run"))
+    # should be test only really:
+    depends_on("py-lxml", type=("run"))
 
     depends_on("darshan-util", type=("build", "run"))
``` 

Now, all the tests pass:

`318 passed, 60 skipped, 17 warnings in 82.71s (0:01:22)`

As a next step--I did my tests with Python `3.10`; I'd suggest repeating my test procedure with Python below version `3.9` to check on the `importlib_resources` issue I mention as well (probably version `3.8`, since that would miss it in the current spackage I think).

If you want to be really thorough, though it is probably overkill for now, to avoid all the test skips you'd need to install the test data files from https://github.com/darshan-hpc/darshan-logs, but that seems like something for a follow-up perhaps.

Reply:
@tylerjereddy and @shanedsnyder I made some changes, and it seems that all the tests are passing when installing `spack install --test=root py-darshan`

Reply:
Oh, I missed your last comments, let me go over them.

Reply:
@tylerjereddy yep, so the changes you suggested were those I made in the previous PR, which fixed the issues and passed all the tests (now using the correct tests). I also tried using Python 3.6 and 3.9, and both worked as expected with no test errors. With Python 3.10, I got the same error reported in #32435 while spack tried to install a dependency. @adamjstewart, do you have any updates on that issue? Other than that, we should be good to go.

Reply:
No updates from me.